api/envoy/api/v3alpha/cds.proto

```diff
@@ -225,6 +230,18 @@ message Cluster {
       // metadata.
       LbSubsetSelectorFallbackPolicy fallback_policy = 2
           [(validate.rules).enum = {defined_only: true}];
+
+      // Subset of
+      // :ref:`keys<envoy_api_field_api.v3alpha.Cluster.LbSubsetConfig.LbSubsetSelector.keys>` used
+      // by
+      // :ref:`KEYS_SUBSET<envoy_api_enum_value_api.v3alpha.Cluster.LbSubsetConfig.LbSubsetSelector.LbSubsetSelectorFallbackPolicy.KEYS_SUBSET>`
+      // fallback policy.
+      // It has to be a non empty list if KEYS_SUBSET fallback policy is selected.
+      // For any other fallback policy the parameter is not used and should not be set.
+      // Only values also present in
+      // :ref:`keys<envoy_api_field_api.v3alpha.Cluster.LbSubsetConfig.LbSubsetSelector.keys>` are
+      // allowed, but `fallback_keys_subset` cannot be equal to `keys`.
```

Comment:
Can you elaborate on these constraints? Why not equal? Why subset? 

What are the implications for performance and memory use during subset construction?

Reply:
> Why not equal?

That constraint is important. If the `fallback_keys_subset` set is not smaller than the `keys` set, we might end up with infinite recursion here <https://github.com/envoyproxy/envoy/pull/8890/files#diff-5d52e6425da0cdb53f84da8906c0a968R256>.
With this constraint we are sure, that in every recursive call the metadata set is smaller than in previous step. If the size of the metadata will be equal to 0, the recursion will inevitably end.

> Why subset?

I assume that you mean "why fallback_keys_subset has to be a subset of keys".
This constraint is actually not important from the technical point of view. If the fallback_keys_subset wouldn't be a subset of keys, it would work fine too.

It rather results from the concept. The concept of this feature is to allow to have "optional" metadata keys in a subset selector. For example: we have a selector for keys ["tag", "canary"], but "canary" is optional and we want to fall back to a selector for keys ["tag"] if there is no host with matching ["tag", "canary"] values.

Without this constraint, we would allow for fallbacks to totally unrelated selectors, for example from ["canary", "tag"] to ["version"]. Maybe it would make sense for someone, I don't know. I can get rid of this constraint, LMK what you think. Also LMK if you have an idea for a name for the fallback policy, because `KEYS_SUBSET` wouldn't fit anymore. 

If we remove that constraint, we can also remove the validation code: <https://github.com/envoyproxy/envoy/pull/8890/files#diff-677be6c88548defd837ae0ddc49468c1R791>. Maybe it would have positive impact for performance during subset construction, but I doubt it would make significant difference.

> What are the implications for performance and memory use during subset construction?

As for memory, we have one additional field in `SubsetSelector`, which is a set that will be empty for a fallback policy other than `KEYS_SUBSET` and one additional pointer in `SubsetSelectorMap`. Number of `SubsetSelector` objects is equal to number of selectors defined in config. Number of `SubsetSelectorMap` objects is I think proportional to sum of a number of unique keys in all selectors.

I tested it in real life scenario, with 1008 EDS clusters and 25192 active subsets in total. I don't see any increase of memory usage when `KEYS_SUBSET` policy is not used. I compared version `v1.12.0` of Envoy to version from this PR. Memory usage (RSS) was in range 270-290 MB for both versions. I deployed both versions alternately multiple times and I didn't observe any reproducible difference. The results are for concurrency=1.

I didn't test it with `KEYS_SUBSET` policy enabled yet, because it would require changes in our Control Plane. But I am not sure if such test would be useful, because memory footprint will vary depending on the `keys` and `fallback_keys_set` size, selectors configuration, etc.

As for performance, I think there is no costly operations introduced during subset creation. I think the most expensive is the validation of fallback_keys_subset for every selector and one set copying per selector (in most cases it will be a very small set). But I am not an C++ expert, LMK if there is a better way to examine performance impact.

Reply:
Thanks. I'll defer to @zuercher on whether the constraints on `fallback_keys_subset` make sense. You have a lot of interesting insight in your comment, it might be helpful to put some summary of this in code or API comments as well for posterity.

Reply:
I just added more comments in code and more information in API docs.

Reply:
I think the constraint makes sense. Enforcing it allows us to be certain there's no possibility of a loop (if e.g. [a,b] falls back to [c] which falls back to [a,b]). If there's need to allow the fallback keys to reference a set of keys that's not a subset of the original we could allow that in a future revision and add some kind of graph traversal to detect loops.